exec/java-exec/src/main/java/org/apache/drill/exec/planner/cost/DrillRelMdDistinctRowCount.java

```diff
@@ -43,6 +48,30 @@ public Double getDistinctRowCount(RelNode rel, ImmutableBitSet groupKey, RexNode
     }
   }
 
+  @Override
+  public Double getDistinctRowCount(Join rel, ImmutableBitSet groupKey, RexNode predicate) {
+    Double count = null;
+    if (rel != null) {
+      if (rel instanceof JoinPrel) {
+        // for Drill physical joins, don't recompute the distinct row count since it was already done
+        // during logical planning; retrieve the cached value.
+        count = ((JoinPrel)rel).getDistinctRowCount();
+        if (count.doubleValue() < 0) {
+          logger.warn("Invalid cached distinct row count for {}; recomputing..", rel.getDescription());
+          count = super.getDistinctRowCount(rel, groupKey, predicate);
+        }
+      } else {
+        count = super.getDistinctRowCount(rel, groupKey, predicate);
```

Comment:
The API of RelMdDistinctRowCount seems to indicate the distinct rowcount depends on input of groupKey and predicate. However, the cached value in DrillJoinRel does not differentiate based on groupKey / predicate. Will it cause issue in the cases this getDistinctRowCount() is called multiple times with different groupKey / predicate?